elasticdl/doc/async_sgd_design.md

```diff
@@ -0,0 +1,118 @@
+Staleness value is the difference between `PS_model_version` and `version`:
+
+```
+staleness = PS_model_version - version
+```
+
+According to some [researches](https://arxiv.org/abs/1810.03264), this staleness affects the training convergence, and large staleness may result in poor training accuracy. The deeper the model, the more impact of the staleness. Some optimizers such as [SGD](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/optimizers/SGD) and [Adagrad](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/optimizers/Adagrad) are more robust to staleness, some optimizers such as other with momentum are very bad with staleness.
+
+[Staleness-aware asychronous SGD](https://arxiv.org/abs/1511.05950) proposes a method to modulate learning rate by the staleness. If the staleness if not 0, this method modulates the learning rate used in the optimizer as:
+
+```
+if staleness > 0:
+    learning_rate_used = learning_rate / staleness
+else:
+    learning_rate_used = learning_rate
+```
+
+### Stale Synchronous Parallel (SSP)
+In the pesudocode for the asynchronous SGD worker, the worker pulls model from PS in every minibatch step. [Stale synchronous parallel (SSP) method](https://dl.acm.org/citation.cfm?id=2999748) uses the strategy that the fastest worker can exceed the slowest one within a predefined staleness threshold. SSP can reduce the number of `get_model_from_ps` calls. The worker training process is:
+
+```
+staleness_threshold = predefined_staleness_threshold
+local_model, model_version = get_model_from_ps()
+local_update_count = 0
+for minibatch in training_data:
+    gradients = compute_gradient(local_model, minibatch)
+    report_gradient_to_ps(gradients, model_version)
+    local_update_count += 1
+    if local_update_count >= staleness_threshold:
+        local_model, model_version = get_model_from_ps()
+        local_update_count = 0
+    else:
+        apply_gradient(local_model, gradients)
+```
+Althrough the original SSP method uses this strategy in synchronized SGD, we can also adopt SSP strategy in asynchronized SGD to reduce `get_model_from_ps` calls.
+
+## Support Asynchronous SGD in ElasticDL
+
+### Change in PS
+1. No need to use locks in `GetModel` and `_update_model` in [server.py](../python/master/servicer.py).
+2. No need to accumulate gradients in `ReportGradient` in [server.py](../python/master/servicer.py). `ReportGradient` calls `_update_model` directly.
+3. Users decide if disabling concurrent variable update by set `use_locking` argument in the optimizer.
+4. To support [Staleness-aware asychronous SGD](https://arxiv.org/abs/1511.05950), PS need to modulate the learning rate in the optimizer with the staleness value.
+
+### Change in Worker
+1. No need to retrain with the minibatch data.
+2. To support SSP strategy, the worker pulls the model from PS in every `staleness_threshold` minibatch steps. Also, the worker needs to update the local model with the computed gradients.
```

Comment:
Here we don't need to update the embedding table when updating the local model. This is because we should pull embedding vectors from embedding service every minibatch.

Reply:
Yes, the local model only includes non-embedding variables.